warehouse/query-core/src/main/java/datawave/query/jexl/visitors/ConjunctionEliminationVisitor.java

```diff
@@ -0,0 +1,105 @@
+     *            a query node
+     * @return a re-written query tree for the node
+     */
+    public static <T extends JexlNode> T optimize(T node) {
+        if (node == null) {
+            return null;
+        }
+        
+        // Operate on copy of query tree.
+        T copy = (T) copy(node);
+        
+        // Visit and enforce collapsing redundant nodes within expression.
+        ConjunctionEliminationVisitor visitor = new ConjunctionEliminationVisitor();
+        copy.jjtAccept(visitor, null);
+        
+        return copy;
+    }
+    
+    @Override
+    public Object visit(ASTOrNode node, Object data) {
+        JexlNode left = getFirstNonASTReference(node.jjtGetChild(0));
+        JexlNode right = getFirstNonASTReference(node.jjtGetChild(1));
+        try {
+            if (isConjunction(left) && hasDuplicate(left, right)) {
+                return copyChildAndUpdateParent(node, right);
+            }
+            if (isConjunction(right) && hasDuplicate(right, left)) {
+                return copyChildAndUpdateParent(node, left);
+            }
+        } catch (ParseException e) {
+            if (log.isDebugEnabled()) {
+                log.debug("Failed to parse child node to check for equivalency", e);
+            }
+        }
+        return node;
+    }
+    
+    // Return the first non-wrapped node.
+    private JexlNode getFirstNonASTReference(JexlNode node) {
+        if (node instanceof ASTReference || node instanceof ASTReferenceExpression) {
+            return getFirstNonASTReference(node.jjtGetChild(0));
+        } else {
+            return node;
+        }
+    }
+    
+    // Return whether or not the given node is an AND.
+    private boolean isConjunction(JexlNode node) {
+        return node instanceof ASTAndNode;
+    }
+    
+    // Return true if the conjunction node
+    private boolean hasDuplicate(JexlNode conjunction, JexlNode otherScript) throws ParseException {
+        int totalChildren = conjunction.jjtGetNumChildren();
+        ASTJexlScript script = getScript(otherScript);
+        for (int i = 0; i < totalChildren; i++) {
+            JexlNode child = conjunction.jjtGetChild(i);
+            if (isEquivalent(child, script)) {
+                return true;
+            }
+        }
+        return false;
+    }
+    
+    // Copy the child node and replace the original node with it in the original node's parent.
+    private JexlNode copyChildAndUpdateParent(JexlNode original, JexlNode child) {
+        JexlNode copy = copy(child);
+        copy.image = original.image;
+        replaceChild(original.jjtGetParent(), original, copy);
+        return copy;
+    }
+    
+    // Return whether or not the two JEXL queries are equivalent.
+    private boolean isEquivalent(JexlNode node, ASTJexlScript script) throws ParseException {
+        ASTJexlScript nodeScript = getScript(node);
+        return TreeEqualityVisitor.isEqual(nodeScript, script, new TreeEqualityVisitor.Reason());
```

Comment:
I thought about comparing the query strings similar to `UniqueExpressionTermsVisitor`, but I found that didn't handle the case where duplicate ANDs/ORs with sub-terms in differing order. It might be worth creating a ticket to come up with a more efficient way to check equivalency between two JexlNodes so that someone can eventually tackle that.

Reply:
Not a bad idea. Opened issue here https://github.com/NationalSecurityAgency/datawave/issues/914